core/src/sql/value/idiom_recursion.rs

```diff
@@ -0,0 +1,156 @@
+use crate::{
+	cnf::IDIOM_RECURSION_LIMIT,
+	ctx::Context,
+	dbs::Options,
+	doc::CursorDoc,
+	err::Error,
+	sql::{part::RecursionPlan, Array, Part},
+};
+
+use super::Value;
+use reblessive::tree::Stk;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Recursion<'a> {
+	pub min: u32,
+	pub max: Option<u32>,
+	pub iterated: u32,
+	pub current: &'a Value,
+	pub path: &'a [Part],
+	pub plan: Option<&'a RecursionPlan>,
+}
+
+impl<'a> Recursion<'a> {
+	pub fn with_iterated(self, iterated: u32) -> Self {
+		Self {
+			iterated,
+			..self
+		}
+	}
+
+	pub fn with_current(self, current: &'a Value) -> Self {
+		Self {
+			current,
+			..self
+		}
+	}
+}
+
+// Method used to check if the value
+// inside a recursed idiom path is final
+pub(crate) fn is_final(v: &Value) -> bool {
+	match v {
+		Value::None => true,
+		Value::Null => true,
+		Value::Array(v) => v.is_empty() || v.is_all_none_or_null(),
+		_ => false,
+	}
+}
+
+pub(crate) fn get_final(v: &Value) -> Value {
+	match v {
+		Value::Array(_) => Value::Array(Array(vec![])),
+		Value::Null => Value::Null,
+		_ => Value::None,
+	}
+}
+
+pub(crate) fn clean_iteration(v: Value) -> Value {
+	if let Value::Array(v) = v {
+		Value::from(v.0.into_iter().filter(|v| !is_final(v)).collect::<Vec<Value>>()).flatten()
+	} else {
+		v
+	}
+}
+
+pub(crate) async fn compute_idiom_recursion<'a>(
+	stk: &mut Stk,
+	ctx: &Context,
+	opt: &Options,
+	doc: Option<&CursorDoc>,
+	rec: Recursion<'a>,
+) -> Result<Value, Error> {
```

Comment:
These lifetimes can be omitted, this is currently not an issue but will become a lint in clippy enabled by default in the future as it is today on nightly rust.
```suggestion
pub(crate) async fn compute_idiom_recursion(
	stk: &mut Stk,
	ctx: &Context,
	opt: &Options,
	doc: Option<&CursorDoc>,
	rec: Recursion<'_>,
) -> Result<Value, Error> {
```